collections/max_by.ts

```diff
@@ -0,0 +1,40 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Returns the first element that is the largest value of the given function or undefined if there are no elements.
+ *
+ * Example:
+ *
+ * ```ts
+ * import { maxBy } from "./max_by.ts";
+ * import { assertEquals } from "../testing/asserts.ts";
+ *
+ * const people = [
+ *     { name: 'Anna', age: 34 },
+ *     { name: 'Kim', age: 42 },
+ *     { name: 'John', age: 23 },
+ * ];
+ *
+ * const personWithMaxAge = maxBy(people, i => i.age);
+ *
+ * assertEquals(personWithMaxAge, { name: 'Kim', age: 42 });
+ * ```
+ */
+export function maxBy<T>(
+  collection: Array<T>,
+  selector: ((el: T) => number | string),
+): T | undefined {
+  let max = collection[0];
+  let maxValue = selector(max);
+
+  for (const current of collection) {
+    const currentValue = selector(current);
+
+    if (currentValue > maxValue) {
+      max = current;
+      maxValue = currentValue;
+    }
+  }
+
+  return max;
```

Comment:
You have not covered the `undefined` case - this will actually error on an empty input. You could do

```suggestion
  let max: T | undefined = undefined;
  let maxValue: ReturnType<typeof selector> = undefined;

  for (const current of collection) {
    const currentValue = selector(current);

    if (maxValue === undefined || currentValue > maxValue) {
      max = current;
      maxValue = currentValue;
    }
  }

  return max;
```

Reply:
if by `empty input` you mean `[]` i don't think there is any bug. since `max` variable initial value is `collection[0]` which in case of empty array is `undefined`. there is already a test case for empty array as input. let me know if I'm wrong.

Reply:
See the test case I posted in the other comment - the problem is that you are passing `undefined` into a function that does not expect `undefined`. Typescript did not catch it for you because deno does not have `noUncheckedIndexAccess` turned on.

Reply:
Thanks for the provided test case. I fixed the issue. please review again and let me know if I missed anything. if the changes are approved I will update 'minBy' PR too. @LionC